### ISSUE TITLE
How to get proper type in for loop collection?

### ISSUE THREAD

**Issue by gwhitney** [DanielXMoore/Civet#1004]
How can I annotate this code to type properly?
```
rawReplacements: Record<string, string> := a: 'z', b: 'y', yz: 'a'
replacements: [RegExp, string][] :=
   ([RegExp(pat, 'g'), rep] for pat, rep in rawReplacements)
```
As far as I can see, the loop collection indeed produces a value of type `[RegExp, string][]` (and I need it typed that way for later code). But I get the error:
```
error TS2322: Type '(string | RegExp)[][]' is not assignable to type '[RegExp, string][]'.
  Type '(string | RegExp)[]' is not assignable to type '[RegExp, string]'.
    Target requires 2 element(s) but source may have fewer.
```
Thanks for any help!

**Comment by edemaine** [comment]
This types correctly, though I don't quite understand why `readonly` is necessary when using `as const`:

```coffee
rawReplacements: Record<string, string> := a: 'z', b: 'y', yz: 'a'
replacements: (readonly [RegExp, string])[] :=
   ([RegExp(pat, 'g'), rep] as const for pat, rep in rawReplacements)
```

**Comment by bbrk24** [comment]
`as const` indicates that the value will never change, so it needs to be `readonly`.

The reason this is happening is that TS is inferring that array literal (without `as const`) as `(RegExp | string)[]` rather than `[RegExp, string]`. Adding `as [RegExp, string]` (where @edemaine put `as const`) should work.

**Comment by gwhitney** [comment]
Ah, I was just typing that: 

I can work around this by writing
```
(([RegExp(pat, 'g'), rep] as [RegExp,string]) for pat, rep in rawReplacements)
```
But it would be nice to not need the "as" clause, since the expression `[RegExp(pat, 'g'), rep]` seems pretty patently to have type `[RegExp, string]`.  But now I see, TypeScript doesn't infer tuple types for array expressions because who knows, you might append another RegExp or string onto that element. There's not a notation for specifically a tuple in TypeScript/Civet, is there, other than the explicit type annotation?

**Comment by STRd6** [comment]
I think we should default to `as const` when aggregating tuples but still allow people to override a different type using `as ...`.

**Comment by bbrk24** [comment]
I'm pretty sure `satisfies [unknown, ...unknown[]]` forces it to be a tuple, but that's longer than the `as` clause in this case. See https://github.com/DanielXMoore/Civet/issues/849#issuecomment-1851254735

**Comment by gwhitney** [comment]
Aha, I just discovered that the following almost trivial operation is super useful in this context:
```
function tupleize<Tup extends unknown[]>(...arg: Tup): Tup arg
```
With that, the following works just fine:
```
rawReplacements  := a: 'z', b: 'y', yz: 'a'
replacements: [RegExp, string][] :=
   (tupleize RegExp(pat, 'g'), rep  for pat, rep in rawReplacements)
```
I might even suggest that if aggregating tuples `as const` is too much by default, that you might want to wrap them in tupleize instead, to give more presumptuous type inference that seems typically useful.

**Comment by STRd6** [comment]
Solved by #1066
